diff -ruN linux-5.9.2/arch/x86/kvm/svm/svm.c linux-5.9.2-patch/arch/x86/kvm/svm/svm.c
--- linux-5.9.2/arch/x86/kvm/svm/svm.c	2020-10-29 10:12:22.000000000 +0100
+++ linux-5.9.2-patch/arch/x86/kvm/svm/svm.c	2020-10-30 22:38:08.893100407 +0100
@@ -1041,6 +1041,7 @@
 	svm_set_intercept(svm, INTERCEPT_XSETBV);
 	svm_set_intercept(svm, INTERCEPT_RDPRU);
 	svm_set_intercept(svm, INTERCEPT_RSM);
+	svm_set_intercept(svm, INTERCEPT_RDTSC);
 
 	if (!kvm_mwait_in_guest(svm->vcpu.kvm)) {
 		svm_set_intercept(svm, INTERCEPT_MONITOR);
@@ -2083,6 +2084,8 @@
 
 static int wbinvd_interception(struct vcpu_svm *svm)
 {
+	svm->vcpu.run->exit_reason = 123;
+	
 	return kvm_emulate_wbinvd(&svm->vcpu);
 }
 
@@ -2091,6 +2094,8 @@
 	u64 new_bv = kvm_read_edx_eax(&svm->vcpu);
 	u32 index = kvm_rcx_read(&svm->vcpu);
 
+	svm->vcpu.run->exit_reason = 123;
+
 	if (kvm_set_xcr(&svm->vcpu, index, new_bv) == 0) {
 		return kvm_skip_emulated_instruction(&svm->vcpu);
 	}
@@ -2170,6 +2175,7 @@
 
 static int cpuid_interception(struct vcpu_svm *svm)
 {
+	svm->vcpu.run->exit_reason = 123;
 	return kvm_emulate_cpuid(&svm->vcpu);
 }
 
@@ -2185,6 +2191,8 @@
 
 static int invd_interception(struct vcpu_svm *svm)
 {
+	svm->vcpu.run->exit_reason = 123;
+	
 	/* Treat an INVD instruction as a NOP and just skip it. */
 	return kvm_skip_emulated_instruction(&svm->vcpu);
 }
@@ -2739,6 +2747,24 @@
 	return nop_interception(svm);
 }
 
+static int handle_rdtsc_interception(struct vcpu_svm *svm) 
+{
+	u64 differece;
+	u64 final_time;
+	u64 data;
+	
+	differece = rdtsc() - svm->vcpu.last_exit_start;
+	final_time = svm->vcpu.total_exit_time + differece;
+
+	data = rdtsc() - final_time;
+
+	svm->vcpu.arch.regs[VCPU_REGS_RAX] = data & -1u;
+	svm->vcpu.arch.regs[VCPU_REGS_RDX] = (data >> 32) & -1u;
+
+	svm->vcpu.run->exit_reason = 123;
+	return nop_interception(svm);
+}
+
 static int (*const svm_exit_handlers[])(struct vcpu_svm *svm) = {
 	[SVM_EXIT_READ_CR0]			= cr_interception,
 	[SVM_EXIT_READ_CR3]			= cr_interception,
@@ -2805,6 +2831,7 @@
 	[SVM_EXIT_RSM]                          = rsm_interception,
 	[SVM_EXIT_AVIC_INCOMPLETE_IPI]		= avic_incomplete_ipi_interception,
 	[SVM_EXIT_AVIC_UNACCELERATED_ACCESS]	= avic_unaccelerated_access_interception,
+	[SVM_EXIT_RDTSC]                = handle_rdtsc_interception,
 };
 
 static void dump_vmcb(struct kvm_vcpu *vcpu)
diff -ruN linux-5.9.2/arch/x86/kvm/x86.c linux-5.9.2-patch/arch/x86/kvm/x86.c
--- linux-5.9.2/arch/x86/kvm/x86.c	2020-10-29 10:12:22.000000000 +0100
+++ linux-5.9.2-patch/arch/x86/kvm/x86.c	2020-10-30 21:29:50.866180190 +0100
@@ -3163,6 +3163,9 @@
 
 int kvm_get_msr_common(struct kvm_vcpu *vcpu, struct msr_data *msr_info)
 {
+	u64 differece;
+	u64 final_time;
+	
 	switch (msr_info->index) {
 	case MSR_IA32_PLATFORM_ID:
 	case MSR_IA32_EBL_CR_POWERON:
@@ -3232,10 +3235,12 @@
 		 * operate L1's TSC value to ensure backwards-compatible
 		 * behavior for migration.
 		 */
-		u64 tsc_offset = msr_info->host_initiated ? vcpu->arch.l1_tsc_offset :
-							    vcpu->arch.tsc_offset;
+		differece = rdtsc() - vcpu->last_exit_start;
+		final_time = vcpu->total_exit_time + differece;
 
-		msr_info->data = kvm_scale_tsc(vcpu, rdtsc()) + tsc_offset;
+		msr_info->data = rdtsc() - final_time;
+
+		vcpu->run->exit_reason = 123;
 		break;
 	}
 	case MSR_MTRRcap:
@@ -8368,7 +8373,7 @@
  * exiting to the userspace.  Otherwise, the value will be returned to the
  * userspace.
  */
-static int vcpu_enter_guest(struct kvm_vcpu *vcpu)
+static int vcpu_enter_guest_real(struct kvm_vcpu *vcpu)
 {
 	int r;
 	bool req_int_win =
@@ -8665,6 +8670,24 @@
 	return r;
 }
 
+static int vcpu_enter_guest(struct kvm_vcpu *vcpu) 
+{
+	int result;
+	u64 differece;
+
+	vcpu->last_exit_start = rdtsc();
+
+	result = vcpu_enter_guest_real(vcpu);
+
+	if (vcpu->run->exit_reason == 123) 
+	{
+		differece = rdtsc() - vcpu->last_exit_start;
+		vcpu->total_exit_time += differece;
+	}
+
+	return result;
+}
+
 static inline int vcpu_block(struct kvm *kvm, struct kvm_vcpu *vcpu)
 {
 	if (!kvm_arch_vcpu_runnable(vcpu) &&
diff -ruN linux-5.9.2/include/linux/kvm_host.h linux-5.9.2-patch/include/linux/kvm_host.h
--- linux-5.9.2/include/linux/kvm_host.h	2020-10-29 10:12:22.000000000 +0100
+++ linux-5.9.2-patch/include/linux/kvm_host.h	2020-10-30 21:30:33.539340889 +0100
@@ -286,6 +286,9 @@
 	unsigned int halt_poll_ns;
 	bool valid_wakeup;
 
+	u64 last_exit_start;
+	u64 total_exit_time;
+
 #ifdef CONFIG_HAS_IOMEM
 	int mmio_needed;
 	int mmio_read_completed;
diff -ruN linux-5.9.2/.vscode/settings.json linux-5.9.2-patch/.vscode/settings.json
--- linux-5.9.2/.vscode/settings.json	1970-01-01 01:00:00.000000000 +0100
+++ linux-5.9.2-patch/.vscode/settings.json	2020-10-30 21:29:37.269606033 +0100
@@ -0,0 +1,5 @@
+{
+    "files.associations": {
+        "ioapic.h": "c"
+    }
+}
\ No newline at end of file
